docs: add sphinx replay description (#9903)
lightningnetwork/lnd: docs/SphinxReplayDB.md

1. **Hop Privacy**: Each intermediate node only knows:
   - The previous hop (where the packet came from)
   - The next hop (where to forward the packet)
   - Cannot see the full route or final destination

2. **Sender Privacy**: Only the sender (Alice) knows:
   - The complete route
   - All intermediate nodes
   - The final destination

3. **Receiver Privacy**: The receiver (Eve) only knows:
   - The immediate previous hop (David)
   - Cannot determine the original sender

4. **Onion Encryption**: Each hop peels one layer of encryption, revealing only
   the next hop's information

The detailed mechanics are described in [BOLT 04](https://github.com/lightning/bolts/blob/master/04-onion-routing.md)

## Replay Protection

Replaying (resending) onion packets into the network can compromise the
privacy guarantees promised by the protocol. So it is crucial for node
participants to not forward replayed onion packets to guard the privacy of all
network participants. Compared to the original Sphinx protocol, the Lightning
Network has an improved replay protection in place especially when it comes to
forwarding HTLCs, which are different from Onion Messages introduced later on

> because they lock a payment to the particular onion packet. Therefore, sending

Reviewer: ...because HTLCs lock a payment to the particular onion packet.

The above framing makes it clear IMO

HTLCs packets comes with a cost. Moreover, every HTLC has an expiry date, also
called CLTV (absolute locktime), which prevents the replay of packets that have
already expired. In addition to the CLTV expiry of a packet, every HTLC onion
packet commits to the payment hash in the HMAC of the message (to be precise,
the associated data), so this prevents an attacker from attaching an old onion
packet to a new payment hash, which now also comes with the risk for the
attacker that he does not only have to lock funds when replaying an onion
packet but he also risks that the next node settles the HTLC because it already
knows the preimage of the HTLC. Although the attack comes with a high cost,
Lightning implementations should prevent replayed onion packets from
propagating through the network to safeguard the privacy for every network
participant. An attacker could, for example, in hindsight access onion packets
from several highly connected nodes and re-inject it into the network and

> potentially unblind receivers of payments which already happened in the past.

Reviewer: Is there an explanation required here on how the "unblinding" will happen? Currently it reads like an assertion.

Author: I generalized the unblind scenario here to not go into too much detail here, because the scenario is very theoretical in my opinion.

### Replay Protection in LND

> For that reason the LND lightning implementation implements the reply protection

Reviewer: nit: LND should be sufficient instead of "LND lightning implementation"

of those  replayed onion packets. In LND there are two interchanging DB names
which save information of those onion packets to prevent replays from happening.

> They are called `Sphinx-Replay-DB` or `Decayed-Log-DB`. Currently LND does only

Reviewer: nit: better framing for improved readability...

Currently LND only implements onion messages which are tied to payments (i.e. HTLCs), and as explained above HTLCs expire after absolute lock time expiry. And therefore can be garbage collected because they will not be forwarded by nodes anyway as that would entail a risk of losing funds.

Carol (Hop 2)
    |
    | Knows: Bob → David
    |
    v
David (Hop 3)
    |
    | Knows: Carol → Eve
    |
    v
Eve (Receiver)
    |
    | Knows: From David
    |
    v

Privacy Properties:
- Each hop only knows its immediate neighbors
- Receiver doesn't know the original sender
- Only sender knows the complete route
- Each hop peels one layer of the onion
```

## Privacy Properties

1. **Hop Privacy**: Each intermediate node only knows:
   - The previous hop (where the packet came from)
   - The next hop (where to forward the packet)
   - Cannot see the full route or final destination

2. **Sender Privacy**: Only the sender (Alice) knows:
   - The complete route
   - All intermediate nodes
   - The final destination

3. **Receiver Privacy**: The receiver (Eve) only knows:
   - The immediate previous hop (David)
   - Cannot determine the original sender

4. **Onion Encryption**: Each hop peels one layer of encryption, revealing only
   the next hop's information

The detailed mechanics are described in [BOLT 04](https://github.com/lightning/bolts/blob/master/04-onion-routing.md)

## Replay Protection

Replaying (resending) onion packets into the network can compromise the

> privacy guarantees promised by the protocol. So it is crucial for node

Author: changed it a bit, now describing replaying of onion packets and then followed by an attack sub chapter.

HTLCs packets comes with a cost. Moreover, every HTLC has an expiry date, also
called CLTV (absolute locktime), which prevents the replay of packets that have
already expired. In addition to the CLTV expiry of a packet, every HTLC onion
packet commits to the payment hash in the HMAC of the message (to be precise,
the associated data), so this prevents an attacker from attaching an old onion
packet to a new payment hash, which now also comes with the risk for the
attacker that he does not only have to lock funds when replaying an onion
packet but he also risks that the next node settles the HTLC because it already
knows the preimage of the HTLC. Although the attack comes with a high cost,
Lightning implementations should prevent replayed onion packets from
propagating through the network to safeguard the privacy for every network
participant. For instance, an attacker could retrospectively acquire onion
packets from several well-connected nodes. Re-injecting these packets into the
network and observing their path could lead to receiver unblinding if an old
packet is detected on an incoming channel but never on an outgoing one from a
given node.

> ### Replay Protection in LND

Reviewer 2: it still isnt clear what exactly is stored in the DB

For that reason the LND implements the replay protection of those replayed onion
packets. In LND there are two interchanging DB names which save information of
those onion packets to prevent replays from happening. They are called 
`Sphinx-Replay-DB` or `Decayed-Log-DB`.
Currently LND only implements onion messages which are tied to payments
(i.e. HTLCs), and as explained above HTLCs expire after absolute lock time. And
therefore can be garbage collected because they will not be forwarded by nodes
anyway as that would entail a risk of losing funds.

> Where is the sphinx replay protection stored for the different backends:

Reviewer 2: ```suggestion
Sphinx replay protection storage for the different backends:
```

1. When running LND with the BBolt backend the db is called: `sphinxreplay.db`

> 2. For Postgres the table is called: `decayedlogdb_kv`

Reviewer 2: ```suggestion
2. For Postgres, the table is called: `decayedlogdb_kv`.
```

> 3. For sqlite the table is called `decayedlogdb_kv` and is part of the

Reviewer 2: ```suggestion
3. For SQLite, the table is called `decayedlogdb_kv` and is part of the
```

>    channel.sqlite file.

Reviewer 2: ```suggestion
   `channel.sqlite` file.
```

#### What happens if LND encounters a replayed onion HTLC packet?

> When LND encounters a replay it will reject the HTLC it will not signal a

Reviewer 2: ```suggestion
When LND encounters a replay, it will reject the HTLC and it will not signal a
```

specific error that a replay occurred which would reveal that the node was
indeed part of the route in the onion packet. LND will reject the HTLC and 

> respond with the message `invalid_onion_version` see also

Reviewer 2: ```suggestion
respond with the message `invalid_onion_version`. See
```

knows the preimage of the HTLC. Although the attack comes with a high cost,
Lightning implementations should prevent replayed onion packets from
propagating through the network to safeguard the privacy for every network
participant.

## Replay attack

An attacker could execute a re-injection attack by strategically positioning
well-connected nodes within the network, ensuring they participate in a high
volume of payment forwarding. In this scenario, the attacker's nodes act as
forwarding intermediaries. These malicious nodes would collect passing onion
packets and concurrently monitor network gossip. Upon observing a successful
payment forward, the attacker could re-inject the captured onion message. By
then monitoring channel updates or closures from the neighboring node, the
attacker aims to determine if that channel was part of the payment path.
Furthermore, if the neighboring node, lacking a suitable outgoing channel,
instead settles the payment (i.e., acts as the final recipient), it becomes
highly probable that it was the ultimate receiving node. It's important to note,
however, that such an attack carries significant costs.

### Replay Protection in LND

> For that reason the LND implements the replay protection of those replayed onion

Reviewer: We can remove this line all together from this section..."For that reason the LND implements the replay protection of those replayed onion packets."

Author: done!
